ccm_web/server/src/canvas/canvas.service.ts

```diff
@@ -47,75 +51,136 @@ export class CanvasService {
-        logger.error(`Received unusual status code ${response.status}`)
-        logger.error(`Response body: ${JSON.stringify(response.data, null, 2)}`)
+        logger.error(
+          `Error occurred while making request to Canvas for access token: ${JSON.stringify(error, null, 2)}`
+        )
       }
-    } catch (error) {
-      logger.error(
-        'Error occurred while making request to Canvas for access token: ',
-        JSON.stringify(error, null, 2)
-      )
+      throw new CanvasOAuthAPIError()
     }
-    if (data === undefined) return false
 
     // Create CanvasToken instance for user
-    let tokenCreated = false
+    const expiresAt = CanvasService.calculateExpiresAt(timeOfRequest, data.expires_in)
     try {
       await this.canvasTokenModel.create({
         userId: user.id,
         accessToken: data.access_token,
-        refreshToken: data.refresh_token
+        refreshToken: data.refresh_token,
+        expiresAt: expiresAt.toISOString()
       })
-      tokenCreated = true
       logger.info(`CanvasToken record successfully created for user ${userLoginId}.`)
     } catch (error) {
       logger.error(
-        'Error occurred while writing Canvas token data to the database: ',
-        JSON.stringify(error, null, 2)
+        `Error occurred while writing Canvas token data to the database: ${JSON.stringify(error, null, 2)}`
       )
+      throw new DatabaseError()
     }
-    return tokenCreated
   }
 
   async findToken (userLoginId: string): Promise<CanvasToken | null> {
     const user = await this.userService.findUserByLoginId(userLoginId)
-    if (user === null) throw new Error(`User with login ID ${userLoginId} was not found!`)
-
+    if (user === null) {
+      logger.error(`User ${userLoginId} is not in the database.`)
+      throw new UserNotFoundError(userLoginId)
+    }
     const token = user.canvasToken === undefined ? null : user.canvasToken
     return token
   }
 
+  async refreshToken (token: CanvasToken): Promise<CanvasToken> {
+    const params = {
+      grant_type: 'refresh_token',
+      client_id: this.clientId,
+      client_secret: this.secret,
+      refresh_token: token.refreshToken
+    }
+
+    let data: TokenRefreshResponseBody | undefined
+    const timeOfRequest = new Date()
+
+    try {
+      const response = await this.httpService.post<TokenRefreshResponseBody>(
+        `${this.url}/login/oauth2/token`, params
+      ).toPromise()
+      logger.debug(`Status code: ${response.status}`)
+      data = response.data
+    } catch (error) {
+      if (axios.isAxiosError(error) && error.response !== undefined) {
+        logger.error(`Received unusual status code ${error.response.status}`)
+        logger.error(`Response body: ${JSON.stringify(error.response.data, null, 2)}`)
+      } else {
+        logger.error(
+          `Error occurred while making request to Canvas for access token: ${JSON.stringify(error, null, 2)}`
+        )
+      }
+      throw new CanvasOAuthAPIError()
```

Comment:
When this case occurs it is ending up in with error message in UI as `{"statusCode":500,"message":"Internal server error"}`. I know that in the API service it is configured like that, but couldn't figure out how this case is trigged in the Canvas service Case. 

In Logs this is how it is show
```
ccm_web     | [Nest] 91   - 07/06/2021, 2:33:19 PM   [ExceptionsHandler] An error occurred while requesting or refreshing a Canvas OAuth access token. +100818ms
ccm_web     | CanvasOAuthError: An error occurred while requesting or refreshing a Canvas OAuth access token.
ccm_web     |     at CanvasService.<anonymous> (/base/server/src/canvas/canvas.service.ts:102:13)
ccm_web     |     at Generator.throw (<anonymous>)
ccm_web     |     at rejected (/base/server/src/canvas/canvas.service.ts:18:65)
ccm_web     |     at processTicksAndRejections (internal/process/task_queues.js:95:5)
```

This comment is for clarity purpose, nothing to change

Reply:
When an error is thrown by a controller-initiated process but isn't caught, NestJS catches it using a global exceptiion filter and throws a `500`. Check out the beginning of this section: https://docs.nestjs.com/exception-filters#exception-filters-1

If we want, I think we can do more here so that we can pass more specific error info to the UI, likely by using a custom exception filter. That might involve making these custom errors NestJS `Exception`s rather than `Error`s. I can look into this more when trying to close #137, or do it now if you want me to.

Reply:
I am ok addressing this as part of #137